### PR TITLE
[SPARK-25362][JavaAPI] Replace Spark Optional class with Java Optional

### PR DESCRIPTION
**Pull request by mmolimar**
## What changes were proposed in this pull request?

Previous Spark versions didn't require Java 8 and an ``Optional`` Spark Java API had to be  implemented to support optional values.

Since Spark 2.4 uses Java 8, the ``Optional`` Spark Java API should be removed so that Spark uses the original Java API.

## How was this patch tested?

``OptionalSuite`` class was removed to test Spark Java API ``Optional`` class (this class as well).
Notice that the ``get`` method in the Spark Java API ``Optional`` class throws a ``NullPointerException`` when the value is not set whereas the native Java API ``java.util.Optional`` throws a ``NoSuchElementException``.

### CODE REVIEW

**Comment by vanzin** [comment]
We can't do this until we decide whether the next version will be a major version or not. This breaks backwards compatibility. (See the bug I duplicated your bug to.)

**Comment by srowen** [comment]
Yeah this is the right kind of change but needs to wait for now. Can you update the title?

**Review by srowen — CHANGES_REQUESTED** [review]
@mmolimar we can move forward with this now. Can you put SPARK-25395 in the title instead? that was the original JIRA.

**Comment by mmolimar** [comment]
Updated @srowen 
The PR title already contains SPARK-25395, is that what you're expecting or another PR?

**Comment by srowen** [comment]
Oops, I mean SPARK-25362. SPARK-25395 was a duplicate.

**Review by srowen — APPROVED** [review]

**Comment by mmolimar** [comment]
No problem. Done ;-)

**Comment by srowen** [comment]
Yeah, you'll have to add this to the 3.0 excludes section of project/MimaExcludes:

`ProblemFilters.exclude[MissingClassProblem]("org.apache.spark.api.java.Optional")`

File: project/MimaExcludes.scala [hunk 1]
```diff
@@ -36,6 +36,8 @@ object MimaExcludes {
 
   // Exclude rules for 3.0.x
   lazy val v30excludes = v24excludes ++ Seq(
+    // [SPARK-25362][JavaAPI] Replace Spark Optional class with Java Optional
+    ProblemFilters.exclude[MissingClassProblem]("org.apache.spark.api.java.Optional")
```
Comment (srowen) [inline project/MimaExcludes.scala]: Sorry @mmolimar this needs a rebase; another change just added a line here and it conflicts. Otherwise looking good

Comment (mmolimar) [reply project/MimaExcludes.scala]: No worries. Done ;-)

Comment (srowen) [reply project/MimaExcludes.scala]: Ah, this has to have a comma after it, and the line that was moved to the end needs the comma removed

File: project/MimaExcludes.scala [hunk 2]
```diff
@@ -36,9 +36,11 @@ object MimaExcludes {
 
   // Exclude rules for 3.0.x
   lazy val v30excludes = v24excludes ++ Seq(
-    ProblemFilters.exclude[DirectMissingMethodProblem]("org.apache.spark.io.SnappyCompressionCodec.version"),
+    // [SPARK-25362][JavaAPI] Replace Spark Optional class with Java Optional
+    ProblemFilters.exclude[MissingClassProblem]("org.apache.spark.api.java.Optional")
     ProblemFilters.exclude[IncompatibleMethTypeProblem]("org.apache.spark.api.java.JavaPairRDD.flatMapValues"),
```
Comment (srowen) [inline project/MimaExcludes.scala]: I realize that I neglected to leave comments in the lines above about the source of these exclusions in earlier changes. While you're here, could you add above this line: 

`// [SPARK-19287] JavaPairRDD flatMapValues requires function returning Iterable, not Iterator`

File: project/MimaExcludes.scala [hunk 3]
```diff
@@ -36,9 +36,11 @@ object MimaExcludes {
...
     ProblemFilters.exclude[IncompatibleMethTypeProblem]("org.apache.spark.streaming.api.java.JavaPairDStream.flatMapValues")
+    ProblemFilters.exclude[DirectMissingMethodProblem]("org.apache.spark.io.SnappyCompressionCodec.version"),
```
Comment (srowen) [inline project/MimaExcludes.scala]: And above the line here, add
`// [SPARK-24109] Remove class SnappyOutputStreamWrapper`

**Comment by srowen** [comment]
Oh, hm:

```
Serialization stack:
	- object not serializable (class: java.util.Optional, value: Optional[x])
	- field (class: scala.Tuple2, name: _2, type: class java.lang.Object)
	- object (class scala.Tuple2, (1,Optional[x]))
	- field (class: scala.Tuple2, name: _2, type: class java.lang.Object)
	- object (class scala.Tuple2, (1,(1,Optional[x])))
	- element of array (index: 0)
	- array (class [Lscala.Tuple2;, size 5)
```

So `java.util.Optional` isn't `Serializable`. Well, that may scuttle this whole idea. I think we're going to find a number of instances where Spark or user apps need to `collect()` `Optional` objects.

Unless someone has a bright idea I think we can't do this. Same reason I was unable to change Spark to use `java.util.function` interfaces -- lambdas aren't otherwise `Serializable` in Java!

**Comment by mmolimar** [comment]
I agree @srowen.
What do you think about reusing the current implementation we already have, for example, in the guava lib instead of having that class in Spark?

**Comment by vanzin** [comment]
Spark's class only reason to exist is so we do *not* use the Guava class, since everyone and their pet want to use a different version of Guava (and Spark shades it for that reason).

**Comment by srowen** [comment]
Yeah, I think we just can't do this unfortunately. It was worth looking into.

**Pull request closed without merging**